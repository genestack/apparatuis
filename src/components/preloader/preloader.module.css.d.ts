/*
 * Copyright (c) 2011-2019 Genestack Limited
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF GENESTACK LIMITED
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 */

export const width100: string;

export const width75: string;

export const width85: string;

export const preloader: string;

export const backgroundRegular: string;

export const backgroundInverted: string;
