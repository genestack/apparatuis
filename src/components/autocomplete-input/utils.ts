/*
 * Copyright (c) 2011-2018 Genestack Limited
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF GENESTACK LIMITED
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 */

// TODO: replace with lodash.debounce
function debounce(func, wait, immediate) { // source: https://gist.github.com/sagiavinash/5c9084b79f68553c4b7d
    let timeout;
    return function(...args:any[]) {
	var context = this;//, args = arguments;
	var later = function() {
	    timeout = null;
	    if (!immediate) func.apply(context, args);
	};
	var callNow = immediate && !timeout;
	clearTimeout(timeout);
	timeout = setTimeout(later, wait);
	if (callNow) func.apply(context, args);
    };
}

function calcMenuStyles(inputDOMNode) {
    if (!inputDOMNode) return {};
    const inputDOMRect = inputDOMNode.getBoundingClientRect();
    return {
        position: 'absolute',
        zIndex: 1070,
        top: inputDOMNode.offsetHeight + inputDOMRect.y + window.pageYOffset + 'px',
        minWidth: inputDOMNode.offsetWidth + 'px',
        left: inputDOMRect.x + window.pageXOffset + 'px'
    };
}

export {
    debounce,
    calcMenuStyles
};
