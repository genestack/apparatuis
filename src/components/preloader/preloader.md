```js
const {DarkContext} = require('../../utils/dark-context');
const {Checkbox} = require('../checkbox');

initialState = {
    contextInverted: false,
    inverted: false,
    drawerOpen: false,
    isLoading: false
};

handleLeftDrawerOpen = () => {
    setState({
        isLoading: true,
        drawerOpen: true
    });
    setTimeout(() => {
        setState({
            isLoading: false
        });
    }, 5000);
};

handleDrawerClose = () =>
    setState({
        isLoading: false,
        drawerOpen: false
    });

<React.Fragment>
    <DarkContext.Provider value={state.contextInverted}>
        <PageContent as={Paper}>
            <Typography variant="section">In context</Typography>
            <Checkbox
                onValueChange={(value) => setState({contextInverted: value})}
                checked={state.contextInverted}
            >
                <Typography>Inverted</Typography>
            </Checkbox>
            <Preloader show count={5} box="paragraph" />
        </PageContent>
    </DarkContext.Provider>

    <DarkContext.Provider value={state.inverted}>
        <PageContent as={Paper}>
            <Typography inverted={state.inverted} variant="section">
                Out of context
            </Typography>
            <Checkbox
                onValueChange={(value) => setState({inverted: value})}
                checked={state.inverted}
            >
                <Typography inverted={state.inverted}>Inverted</Typography>
            </Checkbox>
            <Preloader show count={5} box="paragraph" inverted={state.inverted} />
        </PageContent>
    </DarkContext.Provider>

    <PageContent as={Paper}>
        <Controls justify="space-between">
            <ControlsItem>
                <Button onClick={handleLeftDrawerOpen}>Open Left Drawer</Button>
            </ControlsItem>
        </Controls>
    </PageContent>
    <Drawer open={state.drawerOpen} onClose={handleDrawerClose} side="left">
        <Typography variant="title">Shortcuts</Typography>

        <DrawerFullWidth>
            <Preloader show={state.isLoading} count={7} wrapAll={List} wrapEach={ListItem}>
                <List>
                    <ListItem interactive onClick={handleDrawerClose}>
                        Dashboard
                    </ListItem>
                    <ListItem interactive onClick={handleDrawerClose}>
                        Data Browser
                    </ListItem>
                    <ListItem interactive onClick={handleDrawerClose}>
                        File Manager
                    </ListItem>
                    <ListItem interactive onClick={handleDrawerClose}>
                        Import Data
                    </ListItem>
                </List>

                <List>
                    <ListItem interactive onClick={handleDrawerClose}>
                        Manage Applications
                    </ListItem>
                    <ListItem interactive onClick={handleDrawerClose}>
                        Manage Groups
                    </ListItem>
                </List>
                <Divider />

                <List>
                    <ListItem interactive onClick={handleDrawerClose}>
                        Expression Data Miner
                    </ListItem>
                </List>
            </Preloader>
        </DrawerFullWidth>
        <FlexExpander />

        <Typography variant="caption" quiet box="paragraph">
            Genestack Limited
        </Typography>
        <Typography variant="caption" quiet box="paragraph">
            Copyright (c) 2011-{new Date().getFullYear()}
        </Typography>
    </Drawer>
</React.Fragment>;
```
