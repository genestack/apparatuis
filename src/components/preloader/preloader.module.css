/*
 * Copyright (c) 2011-2019 Genestack Limited
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF GENESTACK LIMITED
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 */

.width100 {
    width: 100%;
}

.width75 {
    width: 75%;
}

.width85 {
    width: 85%;
}

.preloader {
    border-radius: 4px;
}

/**
 * Thin space inserted here to make browser think that there is some content
 * in the Typography component and make It take place
 */
.preloader:before {
    content: "\2009";
    visibility: hidden;
}

.backgroundRegular {
    background-color: rgba(228, 229, 233, 1);
}

.backgroundInverted {
    background-color: rgba(80, 88, 104, 1);
}
