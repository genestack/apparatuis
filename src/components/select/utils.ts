/*
 * Copyright (c) 2011-2020 Genestack Limited
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF GENESTACK LIMITED
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 */

import React from 'react';

import {Option, OptionProps} from './';

// FIXME use selectContext instead this after fixing Menu (Menu must support the "keepMounted" property)
/** finds Option, skips fragments and other react elements */
function findSelectedOption(
    value: number | string,
    children: React.ReactNode
): React.ReactElement<OptionProps> | null {
    let selectedChild = null;

    for (const child of React.Children.toArray(children)) {
        if (selectedChild) {
            return selectedChild;
        }

        if (!React.isValidElement(child)) {
            return null;
        }

        if ((child.type === Option || child.type === 'option') && child.props.value === value) {
            return child as React.ReactElement<OptionProps>;
        }

        selectedChild = findSelectedOption(value, child.props.children);
    }

    return selectedChild;
}

/** Get select label from Options */
export const getSelectLabel = (children: React.ReactNode, selectValue?: number | string) => {
    if (!selectValue) {
        return null;
    }

    const selectedOption = findSelectedOption(selectValue, children);
    if (!selectedOption) {
        return null;
    }

    const {label, value} = selectedOption.props;

    return label ?? value ?? null;
};
