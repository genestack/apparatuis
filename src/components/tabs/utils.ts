/*
 * Copyright (c) 2011-2020 Genestack Limited
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF GENESTACK LIMITED
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 */

import {IndicatorPosition} from '../tab';

import {Variant} from './common-tabs-props';

/** Props of position and size */
export interface PositionStyles {
    top: number;
    left: number;
    width: number;
    height: number;
}

/** Calculation of size and position of indicator */
export function getPositionStyles(
    tabListNode: HTMLElement,
    selectedTabNode: Element,
    indicatorPosition: IndicatorPosition,
    variant: Variant
): PositionStyles {
    const wrapperClientRect = tabListNode.getBoundingClientRect();
    const tabClientRect = selectedTabNode.getBoundingClientRect();

    const top = tabClientRect.top - wrapperClientRect.top;
    const left = tabClientRect.left - wrapperClientRect.left;

    const indicatorHeight = variant === 'ghost' ? 2 : tabClientRect.height;
    const indicatorWidth = variant === 'ghost' ? 2 : tabClientRect.width;

    switch (indicatorPosition) {
        case 'bottom':
            return {
                left,
                top: top + tabClientRect.height - indicatorHeight,
                width: tabClientRect.width,
                height: indicatorHeight
            };
        case 'top':
            return {
                left,
                top,
                width: tabClientRect.width,
                height: indicatorHeight
            };
        case 'right':
            return {
                top,
                left: left + tabClientRect.width - indicatorWidth,
                height: tabClientRect.height,
                width: indicatorWidth
            };
        case 'left':
            return {
                top,
                left,
                height: tabClientRect.height,
                width: indicatorWidth
            };
        default:
            throw new Error('Invalid value for tab indicator position');
    }
}
