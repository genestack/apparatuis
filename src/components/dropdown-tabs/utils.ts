/*
 * Copyright (c) 2011-2020 Genestack Limited
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF GENESTACK LIMITED
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 */

import React, {ReactNode} from 'react';

import {OverridableProps} from '../../utils';
import {TabProps} from '../tab';

interface TabsInfo {
    tabs: React.ReactElement[];
    selectedTabData: OverridableProps<{props: TabProps; defaultType: 'button'}> | null;
}

/** Return valid tabs and selected tab data */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const getTabsInfo = (selectedValue: any, children: ReactNode): TabsInfo => {
    let tabIndex = 0;

    return React.Children.toArray(children).reduce<TabsInfo>(
        (acc, child) => {
            if (!React.isValidElement(child)) {
                return acc;
            }

            const {value, ...restProps} = child.props;
            const childValue = value ?? tabIndex;
            const isActive = childValue === selectedValue;
            tabIndex += 1;

            return {
                tabs: [...acc.tabs, child],
                selectedTabData: isActive
                    ? {value: childValue, children, ...restProps}
                    : acc.selectedTabData
            };
        },
        {
            tabs: [],
            selectedTabData: null
        }
    );
};
