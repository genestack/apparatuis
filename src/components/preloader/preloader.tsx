/*
 * Copyright (c) 2011-2019 Genestack Limited
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF GENESTACK LIMITED
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 */

import classNames from 'classnames';
import * as React from 'react';

import {DarkContext} from '../../utils/dark-context';
import {Typography, TypographyProps} from '../typography';

import * as styles from './preloader.module.css';

/** Preloader public properties */
export interface Props {
    /**
     * How many lines of preloaders should be rendered.
     *
     * Default: `1`
     */
    count?: number;
    /**
     * Custom classnames that should be applied to preloader
     * one by one.
     * Please note, default classnames are imported from the
     * css module.
     *
     * Default: `['width100', 'width75', 'width85']`
     */
    iterateClassnames?: string[];
    /**
     * Describes should preloader be rendered or not.
     * Children will be rendered otherwise.
     *
     * Default `false`
     */
    show?: boolean;
    /**
     * What should be rendered when `show` prop equals to `false`
     *
     * Default `undefined`
     */
    children?: React.ReactNode;
    /**
     * Wrapper for preloader component
     *
     * Default `React.Fragment`
     */
    wrapAll?: React.ElementType;
    /**
     * Wrapper for each preloader line
     *
     * Default: `React.Fragment`
     */
    wrapEach?: React.ElementType;
    /**
     * Typography `className` prop
     */
    className?: TypographyProps['className'];
    /**
     * Typography `inverted` prop
     */
    inverted?: TypographyProps['inverted'];
    /**
     * Typography `variant` prop
     */
    variant?: TypographyProps['variant'];
    /**
     * Typography `box` prop
     */
    box?: TypographyProps['box'];
}

const getIterator = function* <T extends any = any>(array: T[]): IterableIterator<T> {
    while (true) {
        for (const value of array) {
            yield value;
        }
    }
};

const reduceNumber = <T extends any = any>(
    count: number,
    reducer: (value: T, i: number) => T,
    initialValue: T
) => {
    let value = initialValue;
    for (let i = 0; i < count; i += 1) {
        value = reducer(value, i);
    }

    return value;
};

const DEFAULT_CLASSES = [styles.width100, styles.width75, styles.width85];

/**
 * Component that renders preloader lines
 */
export const Preloader = ({
    count = 1,
    iterateClassnames = DEFAULT_CLASSES,
    show = false,
    wrapAll = React.Fragment,
    wrapEach = React.Fragment,
    children,
    className,
    inverted,
    variant,
    box
}: Props) => {
    const classNameGenerator = getIterator(iterateClassnames);
    const ListWrapper = wrapAll;
    const ListItemWrapper = wrapEach;

    if (!show) {
        // Wrapping is required for consistency of returning value
        return <>{children}</>;
    }

    return (
        <ListWrapper>
            {reduceNumber<JSX.Element[]>(
                count,
                (preloaderLines, index) => {
                    const nextClassName = classNameGenerator.next().value;

                    return [
                        ...preloaderLines,
                        <ListItemWrapper key={`${nextClassName}-${index}`}>
                            <DarkContext.Consumer>
                                {(darkContext) => {
                                    const backgroundInverted =
                                        typeof inverted === 'boolean'
                                            ? inverted
                                            : Boolean(darkContext);

                                    return (
                                        <Typography
                                            variant={variant}
                                            box={box}
                                            className={classNames(
                                                styles.preloader,
                                                {
                                                    [styles.backgroundRegular]: !backgroundInverted,
                                                    [styles.backgroundInverted]: backgroundInverted
                                                },
                                                nextClassName,
                                                className
                                            )}
                                            inverted={inverted}
                                        />
                                    );
                                }}
                            </DarkContext.Consumer>
                        </ListItemWrapper>
                    ];
                },
                []
            )}
        </ListWrapper>
    );
};
