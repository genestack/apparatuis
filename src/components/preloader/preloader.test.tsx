/*
 * Copyright (c) 2011-2019 Genestack Limited
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF GENESTACK LIMITED
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 */
import * as React from 'react';
import {create} from 'react-test-renderer';

import {DarkContext} from '../../utils/dark-context';

import {Preloader} from './preloader';

// tslint:disable no-magic-numbers
describe('Preloader component', () => {
    it('renders correctly with count', () => {
        const tree = create(<Preloader count={5} show />);
        expect(tree).toMatchSnapshot();
    });

    it('renders children when show === false', () => {
        const tree = create(<Preloader count={5}>Foo</Preloader>);
        expect(tree).toMatchSnapshot();
    });

    it('works with DarkContext', () => {
        const tree = create(
            <DarkContext.Provider value>
                <Preloader show count={6}>
                    Foo
                </Preloader>
            </DarkContext.Provider>
        );
        expect(tree).toMatchSnapshot();
    });

    it('works with inverted without DarkContext', () => {
        const tree = create(
            <Preloader show count={6} inverted>
                Foo
            </Preloader>
        );
        expect(tree).toMatchSnapshot();
    });

    it('works with custom classnames', () => {
        const tree = create(<Preloader show count={5} iterateClassnames={['foo', 'bar', 'baz']} />);
        expect(tree).toMatchSnapshot();
    });

    it('works with wrapAll and wrapEach as IntrinsicElements', () => {
        const tree = create(<Preloader show count={5} wrapAll="div" wrapEach="span" />);
        expect(tree).toMatchSnapshot();
    });

    it('works with wrapAll and wrapEach as fucntional components', () => {
        const tree = create(
            <Preloader
                show
                count={5}
                wrapAll={({children}: any) => <div>{children}</div>}
                wrapEach={({children}: any) => <span>{children}</span>}
            />
        );
        expect(tree).toMatchSnapshot();
    });
});
