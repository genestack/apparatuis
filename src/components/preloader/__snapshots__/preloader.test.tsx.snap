// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preloader component renders children when show === false 1`] = `"Foo"`;

exports[`Preloader component renders correctly with count 1`] = `
[
  <p
    className="preloader backgroundRegular width100 root body"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundRegular width75 root body"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundRegular width85 root body"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundRegular width100 root body"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundRegular width75 root body"
    data-qa="typography"
  />,
]
`;

exports[`Preloader component works with DarkContext 1`] = `
[
  <p
    className="preloader backgroundInverted width100 root body inverted"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundInverted width75 root body inverted"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundInverted width85 root body inverted"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundInverted width100 root body inverted"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundInverted width75 root body inverted"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundInverted width85 root body inverted"
    data-qa="typography"
  />,
]
`;

exports[`Preloader component works with custom classnames 1`] = `
[
  <p
    className="preloader backgroundRegular foo root body"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundRegular bar root body"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundRegular baz root body"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundRegular foo root body"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundRegular bar root body"
    data-qa="typography"
  />,
]
`;

exports[`Preloader component works with inverted without DarkContext 1`] = `
[
  <p
    className="preloader backgroundInverted width100 root body inverted"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundInverted width75 root body inverted"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundInverted width85 root body inverted"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundInverted width100 root body inverted"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundInverted width75 root body inverted"
    data-qa="typography"
  />,
  <p
    className="preloader backgroundInverted width85 root body inverted"
    data-qa="typography"
  />,
]
`;

exports[`Preloader component works with wrapAll and wrapEach as IntrinsicElements 1`] = `
<div>
  <span>
    <p
      className="preloader backgroundRegular width100 root body"
      data-qa="typography"
    />
  </span>
  <span>
    <p
      className="preloader backgroundRegular width75 root body"
      data-qa="typography"
    />
  </span>
  <span>
    <p
      className="preloader backgroundRegular width85 root body"
      data-qa="typography"
    />
  </span>
  <span>
    <p
      className="preloader backgroundRegular width100 root body"
      data-qa="typography"
    />
  </span>
  <span>
    <p
      className="preloader backgroundRegular width75 root body"
      data-qa="typography"
    />
  </span>
</div>
`;

exports[`Preloader component works with wrapAll and wrapEach as fucntional components 1`] = `
<div>
  <span>
    <p
      className="preloader backgroundRegular width100 root body"
      data-qa="typography"
    />
  </span>
  <span>
    <p
      className="preloader backgroundRegular width75 root body"
      data-qa="typography"
    />
  </span>
  <span>
    <p
      className="preloader backgroundRegular width85 root body"
      data-qa="typography"
    />
  </span>
  <span>
    <p
      className="preloader backgroundRegular width100 root body"
      data-qa="typography"
    />
  </span>
  <span>
    <p
      className="preloader backgroundRegular width75 root body"
      data-qa="typography"
    />
  </span>
</div>
`;
